File: bin/cat_dic.py
"""Concatenate every daily DIC file under dat/ into one CSV for analysis.

Writes incrementally (one daily file at a time) so memory use stays flat
even with thousands of input files.

Output: dat/processed/DOC_Inmates_InCustody_Daily_full.csv
"""

import os
import re
from datetime import datetime

import pandas as pd

OUT_DIR = os.path.join("dat", "processed")
OUT_PATH = os.path.join(OUT_DIR, "DOC_Inmates_InCustody_Daily_full.csv")

# Column names vary across files in case and order, and some files lack
# discharged_dt entirely; normalize to this schema.
COLUMNS = [
    "inmateid", "admitted_dt", "discharged_dt", "custody_level", "bradh",
    "race", "gender", "age", "inmate_status_code", "sealed", "srg_flg",
    "top_charge", "infraction", "date",
]

FILENAME_RE = re.compile(r"DOC_Inmates_InCustody_Daily_(\d{8})\.csv$")


def daily_files():
    for root, dirs, files in os.walk("dat/"):
        # Don't re-ingest our own output.
        dirs[:] = [d for d in dirs if d != "processed"]
        for name in sorted(files):
            match = FILENAME_RE.search(name)
            if match is not None:
                yield os.path.join(root, name), match.group(1)


def main():
    os.makedirs(OUT_DIR, exist_ok=True)
    tmp_path = OUT_PATH + ".tmp"

    n_files = 0
    n_rows = 0
    with open(tmp_path, "w", newline="") as out:
        for path, date_str in daily_files():
            file_date = datetime.strptime(date_str, "%Y%m%d")
            day = pd.read_csv(path, low_memory=False)
            day.columns = day.columns.str.lower()
            day["date"] = file_date
            for col in COLUMNS:
                if col not in day.columns:
                    day[col] = pd.NA
            day[COLUMNS].to_csv(out, index=False, header=(n_files == 0))
            n_files += 1
            n_rows += len(day)
            if n_files % 250 == 0:
                print(f"{n_files} files, {n_rows} rows so far...")

    if n_files == 0:
        os.remove(tmp_path)
        raise SystemExit("no daily files found under dat/")

    os.replace(tmp_path, OUT_PATH)
    print(f"wrote {n_rows} rows from {n_files} files to {OUT_PATH}")


if __name__ == "__main__":
    main()

File: bin/cat_dic.R
library(vroom)
library(tidyverse)
library(stringr)

# Concatenate every daily DIC file under dat/ into dat/combined.csv.

out.path <- "dat/combined.csv"

if (file.exists(out.path)) {
  file.remove(out.path)
}

# Canonical column order; some files lack discharged_dt, and column order
# varies across files, so columns are aligned by name below.
col.order <- c("inmateid", "admitted_dt", "discharged_dt", "custody_level",
               "bradh", "race", "gender", "age", "inmate_status_code",
               "sealed", "srg_flg", "top_charge", "infraction", "date")

data.frame(inmateid = double(), admitted_dt = character(), discharged_dt = character(),
           custody_level = character(), bradh = character(), race = character(),
           gender = character(), age = double(), inmate_status_code = character(),
           sealed = character(), srg_flg = character(), top_charge = character(),
           infraction = character(), date = character()) %>%
  write.table(out.path, sep = ',', row.names = F)

# Only match the daily files themselves, so combined.csv and anything in
# dat/processed/ can't get swept back in.
dic.files <- list.files("dat/", recursive = T,
                        pattern = 'DOC_Inmates_InCustody_Daily_\\d{8}\\.csv$')

n.files <- length(dic.files)
i <- 0

for (file in dic.files) {
  i <- i + 1
  print(paste0(file, " (", i, "/", n.files, ")"))

  date <- as.Date(str_match(file, "(\\d{8})\\.csv$")[[2]], format = '%Y%m%d')
  dic <- vroom(paste0('dat/', file), delim = ',', show_col_types = F, progress = F)
  names(dic) <- tolower(names(dic))
  dic$date <- date

  # Some files do not have a discharged_dt column; add it so every file
  # matches the canonical schema, then align columns by name.
  if (!('discharged_dt' %in% names(dic))) {
    dic$discharged_dt <- NA
  }
  dic <- select(dic, all_of(col.order))

  write.table(dic, out.path, append = T, sep = ',', row.names = F, col.names = F)
}
